securedrop_client/storage.py

```diff
@@ -272,65 +273,69 @@ def update_replies(remote_replies: List[SDKReply], local_replies: List[Reply],
+
+    users: Dict[str, User] = {}
+    for source_uuid, replies in remote_replies_by_source.items():
+        source = session.query(Source).filter_by(uuid=source_uuid).first()
+        for reply in replies:
+            user = users.get(reply.journalist_uuid)
+            if not user:
+                logger.debug(f"user {reply.journalist_uuid} not cached")
+                user = find_or_create_user(
+                    reply.journalist_uuid, reply.journalist_username, session, commit=False
+                )
+                users[reply.journalist_uuid] = user
+
+            local_reply = local_replies_by_uuid.get(reply.uuid)
+            if local_reply:
+                lazy_setattr(local_reply, "journalist_id", user.id)
+                lazy_setattr(local_reply, "size", reply.size)
+                lazy_setattr(local_reply, "filename", reply.filename)
+
+                del local_replies_by_uuid[reply.uuid]
+                logger.debug('Updated reply {}'.format(reply.uuid))
+            else:
+                # A new reply to be added to the database.
+                nr = Reply(uuid=reply.uuid,
+                           journalist_id=user.id,
+                           source_id=source.id,
+                           filename=reply.filename,
+                           size=reply.size)
+                session.add(nr)
+
+                # All replies fetched from the server have succeeded in being sent,
+                # so we should delete the corresponding draft locally if it exists.
+                try:
+                    draft_reply_db_object = session.query(DraftReply).filter_by(
+                        uuid=reply.uuid).one()
+
+                    update_draft_replies(session, draft_reply_db_object.source.id,
+                                         draft_reply_db_object.timestamp,
+                                         draft_reply_db_object.file_counter,
+                                         nr.file_counter, commit=False)
+                    session.delete(draft_reply_db_object)
+
+                except NoResultFound:
+                    pass  # No draft locally stored corresponding to this reply.
+
+                logger.debug('Added new reply {}'.format(reply.uuid))
 
     # The uuids remaining in local_uuids do not exist on the remote server, so
     # delete the related records.
-    replies_to_delete = [r for r in local_replies if r.uuid in local_uuids]
-    for deleted_reply in replies_to_delete:
+    for deleted_reply in local_replies_by_uuid.values():
         delete_single_submission_or_reply_on_disk(deleted_reply, data_dir)
         session.delete(deleted_reply)
         logger.debug('Deleted reply {}'.format(deleted_reply.uuid))
 
     session.commit()
 
 
-def find_or_create_user(uuid: str,
-                        username: str,
-                        session: Session) -> User:
+def find_or_create_user(uuid: str, username: str, session: Session, commit: bool = True) -> User:
```

Comment:
since this is now only used once where commit=False, should we remove the `commit` boolean parameter, or do you think we'll need it elsewhere?

Reply:
The `find_or_create_user` function is also called in `update_and_get_user`, which is called in `Controller.on_authenticate_success`. If no commit is issued there, the database will be locked and prevent syncing. (The functional tests catch this! :tada:)

We could and probably should merge these two functions, and resolve the commit handling, in a separate PR.

Reply:
ah i see, so we could create a new user in the following two cases:

(1) in `update_replies` if we see that a reply has come in from an admin or journalist who doesn't exist yet in our local db
(2) in `on_authenticate_success` if the user logging in exists on the server but not yet locally

in case 2 we want to commit since it's a one-off sign in. in case 1, we could be adding many new users for many new replies if it's the first time we are updating our local db against a server that has many users for instance, so we want to delay a commit until the end of a sync in order to improve performance... mmmkay we can refactor later, but this optimization makes sense.